### PR TITLE
NDRS-608: Add chain_height and current_era metrics

### PR DESCRIPTION
**Pull request by darthsiroftardis**
REF: https://casperlabs.atlassian.net/browse/NDRS-608

Sample output:
`chain_height`
Screenshot from 2020-11-24 09-12-56

`current_era`
Screenshot from 2020-11-24 09-13-29

NOTE: The sample outputs were taken at different instances.

### CODE REVIEW

File: node/src/components/block_executor.rs [hunk 1]
```diff
@@ -159,6 +169,8 @@ impl BlockExecutor {
         // The state hash of the last execute-commit cycle is used as the block's post state
         // hash.
         let next_height = state.finalized_block.height() + 1;
+        // Update the metric.
+        self.metrics.chain_height.set(next_height as i64);
```
Comment (afck) [inline node/src/components/block_executor.rs:173]: Shouldn't we use _this_ block's height instead of the next?

Comment (darthsiroftardis) [reply node/src/components/block_executor.rs:173]: Oh good catch, my mistake, I shall rectify

**Comment by goral09** [comment]
This shouldn't be a metric but information available in the /status endpoint.

**Comment by darthsiroftardis** [comment]
> This shouldn't be a metric but information available in the /status endpoint.
I wouldn't know, I am currently following the story. @EdHastingsCasperLabs opinion?

**Comment by goral09** [comment]
On 20th of October I have finished the task called [NDRS-482: Determine useful consensus metrics](https://casperlabs.atlassian.net/browse/NDRS-483) where I have written down which metrics are useful and what information should be available under `/status` endpoint. I don't know who was the recipient of that analysis though but it looks to me like you could have used it.

**Comment by Fraser999** [comment]
@goral09 @darthsiroftardis This information is indeed already part of the `/status` response, but this ticket was created as a result of a recent request made by Daniel Halford.

I expect our RPC and REST endpoints will continue to evolve for quite some time tbh.

**Comment by goral09** [comment]
Thanks for the update @Fraser999 .

I still don't think this should be part of the metrics. We had a long discussion about what data should go into metrics and what shouldn't. In my opinion this shouldn't.

**Comment by afck** [comment]
I don't really have a strong opinion on this… you're it! :stuck_out_tongue:

**Comment by Fraser999** [comment]
@goral09 
> I still don't think this should be part of the metrics. We had a long discussion about what data should go into metrics and what shouldn't. In my opinion this shouldn't.

I do see your point (I had a similar initial reaction), but I think the benefits of adding this outweigh the negatives.  It's a relatively lightweight change, it won't add a lot of clutter to the outputted metrics, but most importantly it will make our product a lot more convenient to use for one of our external users.  It's not trivial for Daniel to call `/status` by his existing tooling - or rather it's probably more that the two responses use radically different encoding.

**Comment by goral09** [comment]
OK, let's see what Ed says. I argued for similar solutions some time ago and they were turned down.

File: node/src/components/block_executor.rs [hunk 2]
```diff
@@ -40,6 +44,7 @@ use crate::{
     NodeRng,
 };
 pub(crate) use event::Event;
+use prometheus::Registry;
```
Comment (Fraser999) [inline node/src/components/block_executor.rs:47]: This should be with 3rd party imports.

File: node/src/components/consensus/metrics.rs [hunk 3]
```diff
@@ -59,5 +64,8 @@ impl Drop for ConsensusMetrics {
         self.registry
             .unregister(Box::new(self.finalized_block_count.clone()))
             .expect("did not expect deregisterting amount to fail");
```
Comment (Fraser999) [inline node/src/components/consensus/metrics.rs:66]:
```suggestion
            .expect("did not expect deregistering amount to fail");
```

File: node/src/components/consensus/metrics.rs [hunk 4]
```diff
@@ -59,5 +64,8 @@ impl Drop for ConsensusMetrics {
...
+        self.registry
+            .unregister(Box::new(self.current_era.clone()))
+            .expect("did note expect deregistering curren_era to fail");
```
Comment (Fraser999) [inline node/src/components/consensus/metrics.rs:69]:
```suggestion
            .expect("did not expect deregistering current era to fail");
```

**Review by Fraser999 — APPROVED** [review]
Optimistic approval given that the nitpicks are resolved.

**Comment by goral09** [comment]
@Fraser999 I would prefer if you didn't approve the PR, that implements a controversial feature until we get the official agreement. As I said, I argued for this solution months ago and both @EdHastingsCasperLabs and @marc-casperlabs rejected it.

**Comment by Fraser999** [comment]
@goral09 - but I _do_ approve of the PR! :) 
I think the onus is rather on you to block merging by requesting changes or directly asking Karan to not merge until you're satisfied.  With my ageing memory, I'd rather not have to try and remember to come back here to approve something I already reviewed and which I approve of.

**Comment by goral09** [comment]
My point is that we already discussed exactly this feature and decided that we will not implement it via metrics and here's a PR that does it. Anyway, me arguing is probably not worth the time.

File: node/src/components/block_executor.rs (hunk 2 above)
Comment (darthsiroftardis) [reply node/src/components/block_executor.rs:47]: Reordered imports

File: node/src/components/consensus/metrics.rs (hunk 3 above)
Comment (darthsiroftardis) [reply node/src/components/consensus/metrics.rs:66]: Fixed typo

**Pull request merged**